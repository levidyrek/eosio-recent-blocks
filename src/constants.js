/* eslint-disable import/prefer-default-export */
export const SPACE_KEY_CODE = 32;
