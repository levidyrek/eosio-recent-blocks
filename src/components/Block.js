import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { SPACE_KEY_CODE } from '../constants';
import '../stylesheets/Block.css';

export default class Block extends Component {
  initialState = {
    active: false,
  };

  constructor(props) {
    super(props);
    this.state = { ...this.initialState };
  }

  toggleActiveOnKey = e => {
    if (e.which === SPACE_KEY_CODE) {
      e.preventDefault();
      this.toggleActive();
    }
  };

  toggleActive = () => {
    this.setState(state => ({ active: !state.active }));
  };

  render() {
    const { block } = this.props;
    const { active } = this.state;

    const className = active ? 'active' : '';

    return (
      <div key={block.id} className={`block ${className}`}>
        <div
          className="header"
          tabIndex={0}
          role="button"
          onClick={this.toggleActive}
          onKeyDown={this.toggleActiveOnKey}
        >
          <h3>ID: {block.id}</h3>
          <h5>Timestamp: {block.timestamp}</h5>
          <h5>Transactions: {block.num_transactions}</h5>
        </div>
        <div className="raw">
          <h5>Raw content</h5>
          <pre>{block.raw}</pre>
        </div>
      </div>
    );
  }
}

Block.propTypes = {
  block: PropTypes.shape({
    id: PropTypes.string.isRequired,
    num_transactions: PropTypes.number.isRequired,
    raw: PropTypes.string.isRequired,
    timestamp: PropTypes.string.isRequired,
  }).isRequired,
};
