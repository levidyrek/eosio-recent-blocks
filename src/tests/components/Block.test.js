/* eslint-disable no-undef */
import React from 'react';
import renderer from 'react-test-renderer';
import Block from '../../components/Block';
import { SPACE_KEY_CODE } from '../../constants';

const getBlockData = () => ({
  id: '050bd265eba5e0297f2c7e00d299c6f8b492ac184060785fad5576c2afa3e0e7',
  num_transactions: 43,
  raw: '{"timestamp": "2019-10-15T10:48:17.000"}',
  timestamp: '2019-10-15T10:48:17.000',
});

test('Render component', () => {
  const component = renderer.create(<Block block={getBlockData()} />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Toggle active', () => {
  const component = renderer.create(<Block block={getBlockData()} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  component.getInstance().toggleActive();
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

/**
 * Pressing the space bar expands the block.
 */
test('Toggle active on keydown', () => {
  const component = renderer.create(<Block block={getBlockData()} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  const preventDefault = jest.fn();
  const testEvent = { which: SPACE_KEY_CODE, preventDefault };

  component.getInstance().toggleActiveOnKey(testEvent);
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
  expect(preventDefault.mock.calls.length).toBe(1);
});

/**
 * Pressing the wrong key does not expand the block.
 */
test('Toggle active on keydown wrong key', () => {
  const component = renderer.create(<Block block={getBlockData()} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  const preventDefault = jest.fn();
  const testEvent = { which: -1, preventDefault };

  component.getInstance().toggleActiveOnKey(testEvent);
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
  expect(preventDefault.mock.calls.length).toBe(0);
});
