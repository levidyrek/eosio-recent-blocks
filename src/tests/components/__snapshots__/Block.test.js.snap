// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render component 1`] = `
<div
  className="block "
>
  <div
    className="header"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex={0}
  >
    <h3>
      ID: 
      050bd265eba5e0297f2c7e00d299c6f8b492ac184060785fad5576c2afa3e0e7
    </h3>
    <h5>
      Timestamp: 
      2019-10-15T10:48:17.000
    </h5>
    <h5>
      Transactions: 
      43
    </h5>
  </div>
  <div
    className="raw"
  >
    <h5>
      Raw content
    </h5>
    <pre>
      {"timestamp": "2019-10-15T10:48:17.000"}
    </pre>
  </div>
</div>
`;

exports[`Toggle active 1`] = `
<div
  className="block "
>
  <div
    className="header"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex={0}
  >
    <h3>
      ID: 
      050bd265eba5e0297f2c7e00d299c6f8b492ac184060785fad5576c2afa3e0e7
    </h3>
    <h5>
      Timestamp: 
      2019-10-15T10:48:17.000
    </h5>
    <h5>
      Transactions: 
      43
    </h5>
  </div>
  <div
    className="raw"
  >
    <h5>
      Raw content
    </h5>
    <pre>
      {"timestamp": "2019-10-15T10:48:17.000"}
    </pre>
  </div>
</div>
`;

exports[`Toggle active 2`] = `
<div
  className="block active"
>
  <div
    className="header"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex={0}
  >
    <h3>
      ID: 
      050bd265eba5e0297f2c7e00d299c6f8b492ac184060785fad5576c2afa3e0e7
    </h3>
    <h5>
      Timestamp: 
      2019-10-15T10:48:17.000
    </h5>
    <h5>
      Transactions: 
      43
    </h5>
  </div>
  <div
    className="raw"
  >
    <h5>
      Raw content
    </h5>
    <pre>
      {"timestamp": "2019-10-15T10:48:17.000"}
    </pre>
  </div>
</div>
`;

exports[`Toggle active on keydown 1`] = `
<div
  className="block "
>
  <div
    className="header"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex={0}
  >
    <h3>
      ID: 
      050bd265eba5e0297f2c7e00d299c6f8b492ac184060785fad5576c2afa3e0e7
    </h3>
    <h5>
      Timestamp: 
      2019-10-15T10:48:17.000
    </h5>
    <h5>
      Transactions: 
      43
    </h5>
  </div>
  <div
    className="raw"
  >
    <h5>
      Raw content
    </h5>
    <pre>
      {"timestamp": "2019-10-15T10:48:17.000"}
    </pre>
  </div>
</div>
`;

exports[`Toggle active on keydown 2`] = `
<div
  className="block active"
>
  <div
    className="header"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex={0}
  >
    <h3>
      ID: 
      050bd265eba5e0297f2c7e00d299c6f8b492ac184060785fad5576c2afa3e0e7
    </h3>
    <h5>
      Timestamp: 
      2019-10-15T10:48:17.000
    </h5>
    <h5>
      Transactions: 
      43
    </h5>
  </div>
  <div
    className="raw"
  >
    <h5>
      Raw content
    </h5>
    <pre>
      {"timestamp": "2019-10-15T10:48:17.000"}
    </pre>
  </div>
</div>
`;

exports[`Toggle active on keydown wrong key 1`] = `
<div
  className="block "
>
  <div
    className="header"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex={0}
  >
    <h3>
      ID: 
      050bd265eba5e0297f2c7e00d299c6f8b492ac184060785fad5576c2afa3e0e7
    </h3>
    <h5>
      Timestamp: 
      2019-10-15T10:48:17.000
    </h5>
    <h5>
      Transactions: 
      43
    </h5>
  </div>
  <div
    className="raw"
  >
    <h5>
      Raw content
    </h5>
    <pre>
      {"timestamp": "2019-10-15T10:48:17.000"}
    </pre>
  </div>
</div>
`;

exports[`Toggle active on keydown wrong key 2`] = `
<div
  className="block "
>
  <div
    className="header"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex={0}
  >
    <h3>
      ID: 
      050bd265eba5e0297f2c7e00d299c6f8b492ac184060785fad5576c2afa3e0e7
    </h3>
    <h5>
      Timestamp: 
      2019-10-15T10:48:17.000
    </h5>
    <h5>
      Transactions: 
      43
    </h5>
  </div>
  <div
    className="raw"
  >
    <h5>
      Raw content
    </h5>
    <pre>
      {"timestamp": "2019-10-15T10:48:17.000"}
    </pre>
  </div>
</div>
`;
